### PR TITLE
Batching resource state processing; set gRPC message size unlimited; fixed _routers_table.emplace; deleted a lot of loggings for better performance.

### PR DESCRIPTION
**Pull request by zzxgzgz**
This PR does the follow:

- Changed the way ACA processes resource states, introduced batch processing, which enables ACA to process a much higher amount of the same resource state in one GoalState message. Right now 1 million neighbor states (L2 or L3) was tested.
- Added params to ACA's gRPC server to allow unlimited message size, which is very useful when GoalState gets large.
- Changed `_routers_table.emplace(key, value)` to `_routers_table[key] = value`, as we see that values can be missing after using emplace.
- Deleted a lot of loggings, time stamp collecting and elapsed time calculating, as they are slowing down the ACA. Most of the deleted loggings are in the code path of creating neighbors, which has been tested thoroughly.

### CODE REVIEW

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 1]
```diff
@@ -148,14 +149,24 @@ int Aca_Goal_State_Handler::update_port_states(GoalState &parsed_struct,
...
             current_PortState, std::ref(parsed_struct), std::ref(gsOperationReply)));
-
+    if (count % resource_state_processing_batch_size == 0){
+      for (int i = 0; i < workitem_future.size(); i++) {
+        rc = workitem_future[i].get();
+        if (rc != EXIT_SUCCESS)
+          overall_rc = rc;
+      }
+      workitem_future.clear();
+      count = 1;;
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:159]: double ';'

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 2]
```diff
@@ -148,14 +149,24 @@ int Aca_Goal_State_Handler::update_port_states(GoalState &parsed_struct,
...
+    }else{
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:160]: add space please

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 3]
```diff
@@ -180,22 +191,32 @@ int Aca_Goal_State_Handler::update_port_states(GoalStateV2 &parsed_struct,
...
+  int count = 1;
   // below is a c++ 17 feature
   for (auto &[port_id, current_PortState] : parsed_struct.port_states()) {
     ACA_LOG_DEBUG("=====>parsing port state: %s\n", port_id.c_str());
 
     workitem_future.push_back(std::async(
             std::launch::async, &Aca_Goal_State_Handler::update_port_state_workitem_v2, this,
             current_PortState, std::ref(parsed_struct), std::ref(gsOperationReply)));
-
+    if (count % resource_state_processing_batch_size == 0){
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:202]: space

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 4]
```diff
@@ -180,22 +191,32 @@ int Aca_Goal_State_Handler::update_port_states(GoalStateV2 &parsed_struct,
...
+      for (int i = 0; i < workitem_future.size(); i++) {
+        rc = workitem_future[i].get();
+        if (rc != EXIT_SUCCESS)
+          overall_rc = rc;
+      }
+      workitem_future.clear();
+      count = 1;;
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:209]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 5]
```diff
@@ -180,22 +191,32 @@ int Aca_Goal_State_Handler::update_port_states(GoalStateV2 &parsed_struct,
...
+    }else{
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:210]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 6]
```diff
@@ -228,9 +250,20 @@ int Aca_Goal_State_Handler::update_neighbor_states(GoalState &parsed_struct,
             std::launch::async, &Aca_Goal_State_Handler::update_neighbor_state_workitem,
             this, current_NeighborState, std::ref(parsed_struct),
             std::ref(gsOperationReply)));
+    if (count % resource_state_processing_batch_size == 0){
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:253]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 7]
```diff
@@ -228,9 +250,20 @@ int Aca_Goal_State_Handler::update_neighbor_states(GoalState &parsed_struct,
...
+      for (int i = 0; i < workitem_future.size(); i++) {
+        rc = workitem_future[i].get();
+        if (rc != EXIT_SUCCESS)
+          overall_rc = rc;
+      }
+      workitem_future.clear();
+      count = 1;;
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:260]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 8]
```diff
@@ -228,9 +250,20 @@ int Aca_Goal_State_Handler::update_neighbor_states(GoalState &parsed_struct,
...
+    }else{
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:261]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 9]
```diff
@@ -262,9 +296,20 @@ int Aca_Goal_State_Handler::update_router_states(GoalState &parsed_struct,
     workitem_future.push_back(std::async(
             std::launch::async, &Aca_Goal_State_Handler::update_router_state_workitem, this,
             current_RouterState, std::ref(parsed_struct), std::ref(gsOperationReply)));
+    if (count % resource_state_processing_batch_size == 0){
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:299]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 10]
```diff
@@ -262,9 +296,20 @@ int Aca_Goal_State_Handler::update_router_states(GoalState &parsed_struct,
...
+      for (int i = 0; i < workitem_future.size(); i++) {
+        rc = workitem_future[i].get();
+        if (rc != EXIT_SUCCESS)
+          overall_rc = rc;
+      }
+      workitem_future.clear();
+      count = 1;;
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:306]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 11]
```diff
@@ -262,9 +296,20 @@ int Aca_Goal_State_Handler::update_router_states(GoalState &parsed_struct,
...
+    }else{
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:307]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 12]
```diff
@@ -295,9 +341,21 @@ int Aca_Goal_State_Handler::update_neighbor_states(GoalStateV2 &parsed_struct,
             std::launch::async, &Aca_Goal_State_Handler::update_neighbor_state_workitem_v2,
             this, current_NeighborState, std::ref(parsed_struct),
             std::ref(gsOperationReply)));
+    if (count % resource_state_processing_batch_size == 0){
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:344]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 13]
```diff
@@ -295,9 +341,21 @@ int Aca_Goal_State_Handler::update_neighbor_states(GoalStateV2 &parsed_struct,
...
+      for (int i = 0 ; i < workitem_future.size(); i++){
+        rc = workitem_future[i].get();
+        if (rc != EXIT_SUCCESS){
+          overall_rc = rc;
+        }
+      }
+      workitem_future.clear();
+      count = 1;;
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:352]: same as above

File: src/dp_abstraction/aca_goal_state_handler.cpp [hunk 14]
```diff
@@ -295,9 +341,21 @@ int Aca_Goal_State_Handler::update_neighbor_states(GoalStateV2 &parsed_struct,
...
+    }else{
```
Comment (lfu-ps) [inline src/dp_abstraction/aca_goal_state_handler.cpp:353]: same as above

**Review by lfu-ps — COMMENTED** [review]
Removing extra time stamp calculation and loggings look good. 
Please remove the double ';'. 
For the styling, how about run enforcer for these cpp files?

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 1 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 2 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 3 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 4 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 5 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 6 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 7 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 8 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 9 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 10 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 11 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 12 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 13 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

File: src/dp_abstraction/aca_goal_state_handler.cpp (hunk 14 above)
Comment (zzxgzgz) [reply src/dp_abstraction/aca_goal_state_handler.cpp]: fixed

**Comment by zzxgzgz** [comment]
> Removing extra time stamp calculation and loggings look good.
> Please remove the double ';'.
> For the styling, how about run enforcer for these cpp files?

I fixed the spacing and double ';' issue. 

When I used the formatter in my VS Code, it changes the style a lot. That's why I now save files without the formatter. Perhaps we can have a way to format the files in the same style.

**Review by lfu-ps — APPROVED** [review]

**Review by lfu-ps — APPROVED** [review]

**Comment by zzxgzgz** [comment]
The following screenshots shows when testing 1 million neighbors, how long does it take, and how many OVS rules were programmed. 

The mismatch number of rules is likely caused by our test controller, the screenshots emphasize ACA's ability to process a large number of resource states(neighbhor states in this case) and have ovs rules programmed.

image


image

**Pull request merged**